File: extension/lib/memory/chunk.ts
/**
 * Recursive chunker (~400–512 tokens, no overlap) — mirrors the benchmark builder
 * (benchmark/pmrgb/chunk.py) and docs/02 §4. Token count approximated by whitespace.
 */
export interface Chunk {
  chunkId: string;
  text: string;
  offset: [number, number];
}

const ntok = (s: string) => (s.match(/\S+/g)?.length ?? 0);

export function chunkText(text: string, target = 480, hardMax = 512): Chunk[] {
  const paras = text.split('\n').map((p) => p.trim()).filter(Boolean);
  const chunks: Chunk[] = [];
  let buf: string[] = [];
  let bufTok = 0;
  let cursor = 0;

  const flush = () => {
    if (!buf.length) return;
    const body = buf.join('\n');
    let start = text.indexOf(buf[0]!, cursor);
    if (start < 0) start = cursor;
    const end = start + body.length;
    chunks.push({ chunkId: `c_${chunks.length}`, text: body, offset: [start, end] });
    cursor = end;
    buf = [];
    bufTok = 0;
  };

  for (const p of paras) {
    const pt = ntok(p);
    if (bufTok + pt > hardMax && buf.length) flush();
    buf.push(p);
    bufTok += pt;
    if (bufTok >= target) flush();
  }
  flush();
  return chunks.length ? chunks : [{ chunkId: 'c_0', text, offset: [0, text.length] }];
}

File: benchmark/pmrgb/chunk.py
"""Recursive chunker (~400–512 tokens, no overlap) — mirrors docs/02 §4.

Token count approximated by whitespace splitting (good enough for v0; the extension
uses the model tokenizer). Splits on paragraph boundaries first, then packs.
"""
from __future__ import annotations

from .schema import Chunk


def _ntokens(s: str) -> int:
    return len(s.split())


def chunk_text(text: str, target: int = 480, hard_max: int = 512) -> list[Chunk]:
    paras = [p.strip() for p in text.split("\n") if p.strip()]
    chunks: list[Chunk] = []
    buf: list[str] = []
    buf_tok = 0
    cursor = 0  # char offset into `text`

    def flush():
        nonlocal buf, buf_tok, cursor
        if not buf:
            return
        body = "\n".join(buf)
        start = text.find(buf[0], cursor)
        start = start if start >= 0 else cursor
        end = start + len(body)
        chunks.append(Chunk(chunk_id=f"c_{len(chunks)}", text=body, offset=(start, end)))
        cursor = end
        buf, buf_tok = [], 0

    for p in paras:
        pt = _ntokens(p)
        if buf_tok + pt > hard_max and buf:
            flush()
        buf.append(p)
        buf_tok += pt
        if buf_tok >= target:
            flush()
    flush()
    return chunks or [Chunk(chunk_id="c_0", text=text, offset=(0, len(text)))]
